synapse/config/server.py

```diff
@@ -366,11 +369,70 @@ def generate_config_section(
         # Bring DEFAULT_ROOM_VERSION into the local-scope for use in the
         # default config string
         default_room_version = DEFAULT_ROOM_VERSION
+        secure_listeners = []
+        unsecure_listeners = []
+        private_addresses = ["::1", "127.0.0.1"]
+        if listeners:
+            for listener in listeners:
+                if listener["tls"]:
+                    secure_listeners.append(listener)
+                else:
+                    # The open_private_ports option kind of conflicts with the
+                    # idea of passing in your own listener config however
+                    # the local addresses need to be bound if open_private_ports is
+                    # specified.
+                    if open_private_ports:
+                        bind_addresses = listener.setdefault("bind_addresses", [])
+                        for address in private_addresses:
+                            if address not in bind_addresses:
+                                bind_addresses.append(address)
+
+                    unsecure_listeners.append(listener)
+
+            secure_http_bindings = indent(
+                yaml.dump(secure_listeners), " " * 10
+            ).lstrip()
```

Comment:
This indenting looks right, but also feels quite fragile. Can you add some unit tests that set various bits of config and makes sure the yaml round trips through the config generation correctly please?

Reply:
Beauty is in delicate things. So I'll add some inelegant tests.